compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -2283,15 +2283,19 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
             | ty::Error(_)
             | ty::Infer(ty::IntVar(_) | ty::FloatVar(_))
             | ty::Never
+            | ty::Foreign(..)
```

Comment:
Do you expect foreign types to implement any future marker traits?

Reply:
New lang item traits should be implemented for foreign types due to backward compatibility. Example:
```rust

// implicit T: Leak here
fn foo<T: ?Sized>(_: &T) {}

extern "C" {
    type ExternTy;
}

fn forward_extern_ty(x: &ExternTy) {
    foo(x); // ERROR
}

```

It is related to https://github.com/rust-lang/rust/pull/120706#discussion_r1524749883 and same change in an old solver.

For all auto trait other than new lang items foreign types are disqualified earlier at candidate assembly step. https://github.com/rust-lang/rust/blob/30f74ff0dc4d66debc8b50724c446f817e5f75f4/compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs#L710-L715